phoenix-core/src/main/java/org/apache/phoenix/hbase/index/IndexRegionObserver.java

```diff
@@ -917,7 +954,8 @@ private void prepareIndexMutations(BatchMutateContext context, List<IndexMaintai
             for (Pair<IndexMaintainer, HTableInterfaceReference> pair : indexTables) {
                 IndexMaintainer indexMaintainer = pair.getFirst();
                 HTableInterfaceReference hTableInterfaceReference = pair.getSecond();
-                if (nextDataRowState != null) {
+                if (nextDataRowState != null &&
```

Comment:
I think we are not handling the case here when there is an existing row in the index but after an update the new row doesn't satisfy the index where condition so it needs to be deleted from the index table.

Reply:
At-least I don't see any test in the PartialIndexIT which tests this scenario.

Reply:
I will add more tests to cover the case you pointed out.

Reply:
Done